pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/BasePartitionUpsertMetadataManager.java

```diff
@@ -138,15 +138,17 @@ public void addSegment(ImmutableSegment segment) {
       Preconditions.checkState(_enableSnapshot, "Upsert TTL must have snapshot enabled");
       Preconditions.checkState(_comparisonColumns.size() == 1,
           "Upsert TTL does not work with multiple comparison columns");
-      // TODO: Support deletion for TTL. Need to construct queryableDocIds when adding segments out of TTL.
-      Preconditions.checkState(_deleteRecordColumn == null, "Upsert TTL doesn't work with record deletion");
       Number maxComparisonValue =
           (Number) segment.getSegmentMetadata().getColumnMetadataMap().get(_comparisonColumns.get(0)).getMaxValue();
       if (maxComparisonValue.doubleValue() < _largestSeenComparisonValue - _metadataTTL) {
         _logger.info("Skip adding segment: {} because it's out of TTL", segmentName);
         MutableRoaringBitmap validDocIdsSnapshot = immutableSegment.loadValidDocIdsFromSnapshot();
         if (validDocIdsSnapshot != null) {
-          immutableSegment.enableUpsert(this, new ThreadSafeMutableRoaringBitmap(validDocIdsSnapshot), null);
+          ThreadSafeMutableRoaringBitmap queryableDocIds = null;
+          if (_deleteRecordColumn != null) {
+            queryableDocIds = new ThreadSafeMutableRoaringBitmap();
+          }
+          immutableSegment.enableUpsert(this, new ThreadSafeMutableRoaringBitmap(validDocIdsSnapshot), queryableDocIds);
```

Comment:
This is incorrect. queryable doc ids shouldn't be empty. We need to loop over the valid doc ids, find the not deleted records and create the queryable doc ids

Reply:
thanks @Jackie-Jiang i realized the unit test might not cover this function, since we mocked segments and test addSegment(segment, null, null, recordInfoIterator) instead of `addSegment(segment)` after calling enableUpsert().

let me think if there is a better approach to test this part.